### PR TITLE
fix(download): Display errors and stop downloading when server returns an error

### LINKED ISSUE #1575: invalid magic characters 'Inte'
**Describe the bug**
```
 tabby serve --device metal --model TabbyML/StarCoder-1B

Writing to new file.
2024-02-28T21:17:59.085801Z  INFO tabby::serve: crates/tabby/src/serve.rs:116: Starting server, this might take a few minutes...
gguf_init_from_file: invalid magic characters 'Inte'
2024-02-28T21:17:59.270801Z ERROR llama_cpp_bindings: crates/llama-cpp-bindings/src/lib.rs:65: Unable to load model: /Users/nimrod2/.tabby/models/TabbyML/StarCoder-1B/ggml/q8_0.v2.gguf
```
```
cat /Users/nimrod2/.tabby/models/TabbyML/StarCoder-1B/ggml/q8_0.v2.gguf
Internal Error - We're working hard to fix this as soon as possible!
```
**Information about your version**
tabby 0.8.3

**Information about your GPU**
MacOS m1

**Additional context**
follow these instructions
https://tabby.tabbyml.com/docs/installation/apple/

### LINKED ISSUE #1581: "invalid magic characters" if model was downloaded during huggingface maintenance window
**Describe the bug**
During maintenance, huggingface.co models cannot be downloaded. Download attempts lead to the maintenance website being saved to disk instead of the model file.

```
➤ # Start server with StarCoder-1B
> tabby serve --device metal --model TabbyML/DeepseekCoder-6.7B

Writing to new file.
2024-02-28T22:39:07.432479Z  INFO tabby::serve: crates/tabby/src/serve.rs:116: Starting server, this might take a few minutes...
gguf_init_from_file: invalid magic characters '<!DO'
2024-02-28T22:39:07.552864Z ERROR llama_cpp_bindings: crates/llama-cpp-bindings/src/lib.rs:65: Unable to load model: /Users/username/.tabby/models/TabbyML/DeepseekCoder-6.7B/ggml/q8_0.v2.gguf
```

**Information about your version**
Please provide output of `tabby --version`
```
> tabby --version
tabby 0.8.3
```

**Information about your GPU**
Running on Macbook M1.

**Additional context**
grafik

### LINKED ISSUE #1582: [TAB-534] Handle error code (e.g 503) in tabby-download
related issue: [https://github.com/TabbyML/tabby/issues/1575](https://github.com/TabbyML/tabby/issues/1575)

<sub>From [SyncLinear.com](https://synclinear.com) | [TAB-534](https://linear.app/tabbyml/issue/TAB-534/handle-error-code-eg-503-in-tabby-download)</sub>

### PR DESCRIPTION
**Pull request by boxbeam**
Fixes #1582, fixes #1575, fixes #1581, closes TAB-534

### CODE REVIEW

File: crates/aim-downloader/src/https.rs [hunk 1]
```diff
@@ -130,8 +130,15 @@ impl HTTPSHandler {
...
             .send()
             .await
-            .map_err(|_| format!("Failed to GET from {} to {}", &input, &output))
-            .unwrap();
+            .and_then(|r| r.error_for_status())
+            .unwrap_or_else(|e| {
+                eprintln!(
+                    "Failed to download {input}: Server returned {:?}",
+                    e.status().unwrap_or_default()
+                );
```
Comment (wsxiaoys) [inline crates/aim-downloader/src/https.rs:138]: Can this be propagated to application layer (e.g `tabby download` command), thus we can generate better error message there?

Comment (boxbeam) [reply crates/aim-downloader/src/https.rs:138]: What additional context would you want to include?

File: crates/tabby-download/src/lib.rs [hunk 2]
```diff
@@ -65,7 +65,10 @@ async fn download_file(url: &str, path: &Path) -> Result<()> {
 
     let mut bar = WrappedBar::new(0, url, false);
 
-    https::HTTPSHandler::get(url, &intermediate_filename, &mut bar, "").await?;
+    if let Err(e) = https::HTTPSHandler::get(url, &intermediate_filename, &mut bar, "").await {
+        eprintln!("Fetching model failed: {e}");
```
Comment (wsxiaoys) [inline crates/tabby-download/src/lib.rs:69]: use `tracing::error`, generate error message based on `url` and status code

File: crates/tabby-download/src/lib.rs [hunk 3]
```diff
@@ -65,7 +65,10 @@ async fn download_file(url: &str, path: &Path) -> Result<()> {
...
+        std::process::exit(1);
```
Comment (wsxiaoys) [inline crates/tabby-download/src/lib.rs:70]: there's no need to call `exit`, see handler code in line 82

File: crates/aim-downloader/src/https.rs [hunk 4]
```diff
@@ -130,8 +135,12 @@ impl HTTPSHandler {
             .basic_auth(parsed_address.username, Some(parsed_address.password))
             .send()
             .await
-            .map_err(|_| format!("Failed to GET from {} to {}", &input, &output))
-            .unwrap();
+            .and_then(|r| r.error_for_status())
+            .map_err(|e| DownloadError::ServerError {
```
Comment (wsxiaoys) [inline crates/aim-downloader/src/https.rs:139]:
```suggestion
            .map_err(|e| DownloadError::HTTPError {
```

**Review by wsxiaoys — COMMENTED** [review]
Please provide a oneliner to test the behavior, (e.g `serve --model <model_url_return_error_code>`

File: crates/tabby-download/src/lib.rs (hunk 2 above)
Comment (boxbeam) [reply crates/tabby-download/src/lib.rs:69]: `DownloadError` already gets converted into an error message containing the URL and status code

File: crates/tabby-download/src/lib.rs (hunk 3 above)
Comment (boxbeam) [reply crates/tabby-download/src/lib.rs:70]: Not sure whether you want me to use `panic!` or `error!`

**Comment by boxbeam** [comment]
To test: `tabby serve --model boxbeam/StarCoder-1B`

File: crates/tabby-download/src/lib.rs (hunk 2 above)
Comment (wsxiaoys) [reply crates/tabby-download/src/lib.rs]: Let's generate the message here, making it easier to create more fine graind err msg in future

File: crates/tabby-download/src/lib.rs (hunk 3 above)
Comment (wsxiaoys) [reply crates/tabby-download/src/lib.rs]: propagating error object would be good

File: crates/tabby-download/src/lib.rs [hunk 5]
```diff
@@ -65,7 +65,16 @@ async fn download_file(url: &str, path: &Path) -> Result<()> {
 
     let mut bar = WrappedBar::new(0, url, false);
 
-    https::HTTPSHandler::get(url, &intermediate_filename, &mut bar, "").await?;
+    if let Err(e) = https::HTTPSHandler::get(url, &intermediate_filename, &mut bar, "").await {
+        match e {
+            DownloadError::HttpError { name, code } => {
+                bail!("Fetching file {name} failed: Server returned HTTP code {code}")
```
Comment (wsxiaoys) [inline crates/tabby-download/src/lib.rs:71]:
```suggestion
                bail!("Fetching '{name}' failed: Server returned {code} HTTP status")
```

File: crates/tabby-download/src/lib.rs [hunk 6]
```diff
@@ -65,7 +65,16 @@ async fn download_file(url: &str, path: &Path) -> Result<()> {
...
+            }
+            DownloadError::Validate { source } => {
+                bail!("Fetching file {source} failed: File failed to validate")
```
Comment (wsxiaoys) [inline crates/tabby-download/src/lib.rs:74]:
```suggestion
                bail!("Failed to validate '{source}'")
```

**Review by wsxiaoys — APPROVED** [review]
Otherwise LGTM

**Pull request merged**